src/Storages/MergeTree/MergeTreeIndexGin.cpp

```diff
@@ -760,42 +797,117 @@ MergeTreeIndexConditionPtr MergeTreeIndexGin::createIndexCondition(const Actions
 
 MergeTreeIndexPtr ginIndexCreator(const IndexDescription & index)
 {
-    size_t ngram_length = index.arguments.empty() ? 0 : index.arguments[0].safeGet<size_t>();
-    UInt64 max_rows = index.arguments.size() < 2 ? DEFAULT_MAX_ROWS_PER_POSTINGS_LIST : index.arguments[1].safeGet<UInt64>();
-    GinFilterParameters gin_filter_params(ngram_length, max_rows);
-
-    if (ngram_length == 0)
+    const FieldVector & arguments = index.arguments;
+    bool is_key_value_pair = !arguments.empty() && arguments[0].getType() == Field::Types::Tuple;
+    if (is_key_value_pair)
     {
-        auto tokenizer = std::make_unique<SplitTokenExtractor>();
-        return std::make_shared<MergeTreeIndexGin>(index, gin_filter_params, std::move(tokenizer));
+        auto options = index.arguments | std::views::transform([](const auto & argument) { return argument.template safeGet<Tuple>(); });
```

Comment:
(cosmetic: there are only few usages of `std::ranges` in the codebase and these few ones are very simplistic, it would be preferred to do it the old-fashioned way)

Reply:
I removed the usage of `std::ranges` and changed the extract option function to support extraction of a generic type.